codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +70,1047 @@
+    }
+
+    // TODO: Handle writability changes due writing from outside the eventloop.
+    private final class DefaultHttp2StreamChannel extends DefaultAttributeMap implements Http2StreamChannel {
+        private final Http2StreamChannelConfig config = new Http2StreamChannelConfig(this);
+        private final Http2ChannelUnsafe unsafe = new Http2ChannelUnsafe();
+        private final ChannelId channelId;
+        private final ChannelPipeline pipeline;
+        private final Http2FrameStream stream;
+        private final ChannelPromise closePromise;
+
+        private volatile boolean registered;
+        private volatile boolean writable;
+
+        private boolean closePending;
+        private boolean readInProgress;
+        private Queue<Object> inboundBuffer;
+
+        /** {@code true} after the first HEADERS frame has been written **/
+        private boolean firstFrameWritten;
+
+        /** {@code true} if a close without an error was initiated **/
+        private boolean streamClosedWithoutError;
+
+        // Keeps track of flush calls in channelReadComplete(...) and aggregate these.
+        private boolean inFireChannelReadComplete;
+        private boolean flushPending;
+
+        boolean fireChannelReadPending;
+
+        // Holds the reference to the next DefaultHttp2StreamChannel that should be processed in
+        // channelReadComplete(...)
+        DefaultHttp2StreamChannel next;
+
+        DefaultHttp2StreamChannel(Http2FrameStream stream) {
+            this.stream = stream;
+            ((Http2MultiplexCodecStream) stream).channel = this;
+            pipeline = new DefaultChannelPipeline(this) {
+                @Override
+                protected void incrementPendingOutboundBytes(long size) {
+                    // Do thing for now
                 }
-            }
+
+                @Override
+                protected void decrementPendingOutboundBytes(long size) {
+                    // Do thing for now
+                }
+            };
+            closePromise = pipeline.newPromise();
+            channelId = new Http2StreamChannelId(parent().id(), ++idCount);
         }
-    }
 
-    @SuppressWarnings("unchecked")
-    private static void initAttrs(Channel channel, Map<AttributeKey<?>, Object> attrs) {
-        if (attrs != null) {
-            for (Entry<AttributeKey<?>, Object> e: attrs.entrySet()) {
-                channel.attr((AttributeKey<Object>) e.getKey()).set(e.getValue());
+        @Override
+        public Http2FrameStream stream() {
+            return stream;
+        }
+
+        void streamClosed() {
+            streamClosedWithoutError = true;
+            if (readInProgress) {
+                // Just call closeForcibly() as this will take care of fireChannelInactive().
+                unsafe().closeForcibly();
+            } else {
+                closePending = true;
             }
+            //((Http2MultiplexCodecStream) stream).channel = null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)